### PR TITLE
Fix directory table problems and add guc allow_dml_directory_table.

### PR DESCRIPTION
**Pull request by wenchaozhang-123**
In this commit, fix gpdirtable can not specify file name. We can specify file name using two match mode, full match and regex match. Full match using for download one file while regex match using for downloading files in directory. What's more, we add new guc allow_dml_directory_table, which allow us to do dml on directory table's schema table. Lastly, we support a new grammar for drop directory table. We don't remove directory table's directory by default, while we can remove it by using 'with content' keywords in drop directory table grammar, for example:

DROP DIRECTORY TABLE dir_table WITH CONTENT;

Authored-by: Zhang Wenchao zwcpostgres@gmail.com

fix #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

File: gpMgmt/bin/gpdirtableload [hunk 1]
```diff
@@ -355,17 +363,24 @@ class gpdirtableload:
...
+        qry = "SELECT relative_path FROM %s WHERE " % self.options.table
+
+        qry += " relative_path like \'%s" % self.options.input_file
+        if self.regexMatch:
+        	qry += '%'
+        qry += "\'"
 
         if self.options.tag:
-            qry += "WHERE tag = \'%s\'" % self.options.tag
+            qry += ", tag = \'%s\'" % self.options.tag
```
Comment (gfphoenix78) [inline gpMgmt/bin/gpdirtableload:374]: The conditions should be joined by 'AND'. Do we have test cases for these options?

File: src/backend/catalog/dependency.c [hunk 2]
```diff
@@ -1489,12 +1490,22 @@ doDeletion(const ObjectAddress *object, int flags)
 				}
 				else
 				{
+					bool 		drop_with_content = ((flags & PERFORM_DELETION_WITH_CONTENT) != 0);
```
Comment (gfphoenix78) [inline src/backend/catalog/dependency.c:1493]: The outer '()' is unneeded.

File: src/backend/nodes/copyfuncs.c [hunk 3]
```diff
@@ -6751,6 +6779,7 @@ copyObjectImpl(const void *from)
 		case T_DropStmt:
 			retval = _copyDropStmt(from);
 			break;
+			break;
```
Comment (gfphoenix78) [inline src/backend/nodes/copyfuncs.c:6782]: duplicated break

File: gpMgmt/bin/gpdirtableload (hunk 1 above)
Comment (wenchaozhang-123) [reply gpMgmt/bin/gpdirtableload]: Fixed,  as this tools is independent from kernel, does not add test cases now.

File: src/backend/nodes/outfuncs.c [hunk 4]
```diff
@@ -4045,6 +4045,25 @@ _outAlterDirectoryTableStmt(StringInfo str, const AlterDirectoryTableStmt *node)
...
+
+static void
+_outDropDirectoryTableStmt(StringInfo str, const DropDirectoryTableStmt *node)
+{
+	WRITE_NODE_TYPE("DROPDIRECTORYTABLESTMT");
+
+	_outDropStmtInfo(str, (const DropStmt *) node);
+	WRITE_BOOL_FIELD(with_content);
+}
+
```
Comment (gfphoenix78) [inline src/backend/nodes/outfuncs.c:4066]: These non-standard functions could be contained in `outfuncs_common.c`.
It has two benefits:
1. the `out` functions will have text and binary versions.
2. The additional code causes unneeded conflict in kernel upgrade.

File: src/backend/nodes/readfast.c [hunk 5]
```diff
@@ -1774,6 +1774,33 @@ _readAlterDirectoryTableStmt(void)
...
+{
+	READ_LOCALS(DropDirectoryTableStmt);
+
+	_readDropStmt_common(&local_node->base);
+
+	READ_BOOL_FIELD(with_content);
+
+	READ_DONE();
+}
+
```
Comment (gfphoenix78) [inline src/backend/nodes/readfast.c:1803]: Like out functions, these `read` functions could be placed in `readfuncs_common.c`

File: src/backend/nodes/outfuncs.c (hunk 4 above)
Comment (wenchaozhang-123) [reply src/backend/nodes/outfuncs.c:4066]: good idea

**Review by gfphoenix78 — APPROVED** [review]

**Review by gfphoenix78 — APPROVED** [review]

**Review by gongxun0928 — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]

**Pull request merged**